File: Editorial/0105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/JavaScript/1_Approach_Recursion.js
var buildTree = function (preorder, inorder) {
    let preorderIndex = 0;
    let inorderIndexMap = new Map();
    for (let i = 0; i < inorder.length; i++) {
        inorderIndexMap.set(inorder[i], i);
    }
    function arrayToTree(left, right) {
        // if there are no elements to construct the tree
        if (left > right) return null;
        // select the preorder_index element as the root and increment it
        let rootValue = preorder[preorderIndex++];
        let root = new TreeNode(rootValue);
        // build left and right subtree
        // excluding inorderIndexMap[rootValue] element because it's the root
        root.left = arrayToTree(left, inorderIndexMap.get(rootValue) - 1);
        root.right = arrayToTree(inorderIndexMap.get(rootValue) + 1, right);
        return root;
    }
    return arrayToTree(0, preorder.length - 1);
};

File: Editorial/0105_Construct_Binary_Tree_from_Preorder_and_Inorder_Traversal/Go/1_Approach_Recursion.go
func buildTree(preorder []int, inorder []int) *TreeNode {
    preorderIndex := 0
    inorderIndexMap := make(map[int]int)
    for i := 0; i < len(inorder); i++ {
        inorderIndexMap[inorder[i]] = i
    }
    var arrayToTree func(left int, right int) *TreeNode
    arrayToTree = func(left int, right int) *TreeNode {
        // if there are no elements to construct the tree
        if left > right {
            return nil
        }
        // select the preorder_index element as the root and increment it
        rootValue := preorder[preorderIndex]
        preorderIndex++
        root := &TreeNode{Val: rootValue}
        // build left and right subtree
        // excluding inorderIndexMap[rootValue] element because it's the root
        root.Left = arrayToTree(left, inorderIndexMap[rootValue]-1)
        root.Right = arrayToTree(inorderIndexMap[rootValue]+1, right)
        return root
    }
    return arrayToTree(0, len(preorder)-1)
}
